indexer/SymbolFormatter.h

```diff
@@ -72,6 +104,18 @@ class SymbolFormatter final {
   std::optional<std::string_view>
   getSymbolCached(const clang::Decl *,
                   absl::FunctionRef<std::optional<std::string>()>);
+
+  /// Format the string to a buffer stored by `this` and return a view to it.
+  template <typename... T>
```

Comment:
It would be great to set up ASAN to catch cases when we mess up and hold onto the string_view too long.

In general, is this error-prone? You could hold onto the string_view too long. You could format it twice and overwrite the result of the first one.

Could you make the abstraction harder to use incorrectly? Like disallow copy and assignment; hand out a helper object which is operator* dereferenceable to const std::string_view& but discourages you from misadventure? Basically force people to use local variable scope for lifetime and flip a bit when in use and crash early and often?

Reply:
ASAN is already turned on in the local dev build and for CI testing. 

> In general, is this error-prone? You could hold onto the string_view too long. You could format it twice and overwrite the result of the first one.

I agree in the general case. In this specific case, I think the biggest risk is probably formatting it twice. I initially considered having an RAII type similar to what you're describing, which would prevent you from two of these in flight.

However, I think the risk is somewhat reduced because the current API really forces you into the idiom of:
1. Get a context symbol (this is a string_view not into the scratch buffer)
2. Create a descriptor for the entity.
3. Use `formatTemporary` (the scratch buffer is in use)
4. Immediately consume the two string views and make a longer `string` which is stored in the cache.

For symbols where context isn't used/applicable, you'd likely pass arguments to the variadic version once and be done in a single call.

It's unlikely that we need to deviate from this pattern. If we do deviate from this pattern, and the underlying buffer stays the same, then when the code path is exercised in a test (which it should be, if we're not adding a test for a new entity along with the implementation code, then the PR shouldn't be merged), then you'll get incorrect snapshot output and notice that something is wrong.

If the underlying buffer gets reallocated elsewhere, you'll get an ASan failure.

So in this specific case, I think it's unlikely that you'd run into buffer misuse (or run into it, but be able to fix it quickly on exercising the codepath) because of the surrounding circumstances. So I've opted to do the simpler thing here first -- we can add more safeguards later if this API turns out to be too error prone in practice.